### ISSUE TITLE
Problem Getting Token from User Assigned Managed Identity in Azure Container Instance

### ISSUE THREAD

**Issue by AJAnderhub** [Azure/azure-sdk-for-python#25127]
- **Package Name**: Azure.Identity
- **Package Version**: 1.10.0
- **Operating System**: Ubuntu 18.04
- **Python Version**:  3.8.0

**Describe the bug**
I am trying to run a python script in an Azure Container instance with a user assigned managed identity that is using the Azure Identity SDK for Python to get a token from the managed identity endpoint in the container instance. The SDK should be making a request to IMDS endpoint in order to get a token. The Azure Cli (2.37.0) is able to do so, but the Azure.Identity python SDK is not. I get the following error.

2022-07-06 03:35:20.275989: WARNING azure.identity._credentials.chained:chained.py:107 ChainedTokenCredential failed to retrieve a token from the included credentials.
2022-07-06 03:35:20.275994: Attempted credentials:
2022-07-06 03:35:20.276002: ManagedIdentityCredential: Unexpected content type "text/plain; charset=utf-8"
2022-07-06 03:35:20.276008: Content: Received invalid token. Please try again.

I don’t have the exact stack trace since the errors are caught and not propagated up. But it seems to be failing on line 58 [here](https://github.com/Azure/azure-sdk-for-python/blob/12f3e62f781436dfbd1c21e1dd8cb0f5ee3da629/sdk/identity/azure-identity/azure/identity/_internal/managed_identity_client.py) because of expecting the format "application/json" but gets "text/plain; charset=utf-8".

I took a look at what that plaintext response which was: “Required metadata header not specified or not correct”. So, I looked at the request in the SDK and it appears that it is not adding the header “Metadata: true” to the GET request to the IMDS endpoint that is required to get the token. It looks like the headers are supposed to be added in line 97 [here](https://github.com/Azure/azure-sdk-for-python/blob/12f3e62f781436dfbd1c21e1dd8cb0f5ee3da629/sdk/identity/azure-identity/azure/identity/_credentials/imds.py) . But that doesn’t seem to be making it to the final request.

**To Reproduce**
Attempt to get a token from a user assigned managed identity user the azure.identity sdk in an Azure container instance. (the Azure container instance is a product of an internal Ev2 shell extension).

**Expected behavior**
The SDK is able to get a token from the IMDS endpoint.

**Comment by xiangyan99** [comment]
Thanks for the feedback, we’ll investigate asap.

**Comment by AJAnderhub** [comment]
@xiangyan99 the current steps we are taking.
1. Creating a `azure.identity.ChainedTokenCredential` and adding `azure.identity.ManagedIdentityCredential(client_id=self.client_id, **kwargs,)` with the Client ID of the managed identity being used.
2. Calling `chained_credential_object.get_token(scope)` where the `scope` is the Application ID of a registered application in Azure. So, `scope = "00000000-0000-0000-0000-000000000000/.default` (real application ID redacted)

**Comment by xiangyan99** [comment]
Thanks for the information.

Did you assign the identity to the service instance?

**Comment by AJAnderhub** [comment]
Ev2 handles that part [automatically](https://ev2docs.azure.net/features/extensibility/shell/intro.html#managed-identity). But I believe that it is assigned since I was able to retrieve a token using the Azure CLI by running:
```pwsh
az login --identity
az account set --subscription $SubscriptionId
az account get-access-token --resource 00000000-0000-0000-0000-000000000000
```

**Comment by xiangyan99** [comment]
This is what I meant.

![image](https://user-images.githubusercontent.com/14350651/181645659-39c839d0-bfb9-4d1b-99d0-8d22c3a6fa37.png)

I am afraid your script may not assign the identity for your automatically.

**Comment by AJAnderhub** [comment]
Like I mentioned, Ev2 handles the linking of the managed identity to the container instance:
![ACI Managed Identity](https://user-images.githubusercontent.com/37024457/181649019-13c569c3-3aa2-45ab-b6d9-d9d38bd5cdef.png)

**Comment by xiangyan99** [comment]
Thanks for your confirmation.

Let me rephrase to make sure we are on the same page:
You called
```py
identity = azure.identity.ManagedIdentityCredential(client_id=self.client_id)
identity.get_token(scope)
```
and got
```shell
ManagedIdentityCredential: Unexpected content type "text/plain; charset=utf-8"
```

right?

(I removed ChainedTokenCredential and only used ManagedIdentityCredential to make it simpler.)

**Comment by AJAnderhub** [comment]
Correct.

**Comment by xiangyan99** [comment]
Can you enable the log by adding

```py
import logging
logging.basicConfig(filename='c:/temp/log.log', level=logging.DEBUG)
```

and when creating credential object, use
```py
identity = azure.identity.ManagedIdentityCredential(client_id=self.client_id, logging_enable=True)
```
?

Thanks

**Comment by AJAnderhub** [comment]
So, after testing a cut down hard-coded version like you said it worked. I compared the results with our library's original call and it was requesting a token from the wrong resource ID (it was using the resource ID for the configuration of a different tenant). So, that seemed to be the issue. The error message that I first posted with was just misleading for tracking that down.

So I guess this issue can be closed unless changing that error message to something more meaningful is something that would be worth doing.
